File: Problems/0108. (E) Convert Sorted Array to Binary Search Tree/Python3/solution.py
class Solution:
  def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
    # recursive solution
    return self.recurse(nums, 0, len(nums)-1)
  
  def recurse(self, nums, l, h):
    arr_len = h - l
    arr_mid = (l + h) // 2

    if arr_len < 0:
      return None
    
    # recursively generate left and right subtrees
    left = self.recurse(nums, l, arr_mid-1)
    right = self.recurse(nums, arr_mid+1, h)
    cur = TreeNode(nums[arr_mid], left, right)
    return cur

File: Problems/0108. (E) Convert Sorted Array to Binary Search Tree/Java/solution.java
class Solution {
  // recursive solution
  public TreeNode sortedArrayToBST(int[] nums) {
    return recurse(nums, 0, nums.length-1);
  }
  
  private TreeNode recurse(int[] nums, int l, int h) {
    int arr_len = h - l;
    int arr_mid = (l + h) / 2;

    if (arr_len < 0)
      return null;
    
    // recursively generate left and right subtrees
    TreeNode left = recurse(nums, l, arr_mid-1);
    TreeNode right = recurse(nums, arr_mid+1, h);
    return new TreeNode(nums[arr_mid], left, right);
  }
}
